enhancements/automated-resource-request-scaling.md

```diff
@@ -0,0 +1,635 @@
+  fewer than the default resources would offer exactly the same
+  performance as the default. It must be stable and continue to
+  operate reliably, but may respond more slowly.
+* This enhancement does not address scaling the control plane **up**.
+
+## Proposal
+
+There are two aspects to manage scaling the control plane resource
+requests. Most of the cluster operators are deployed by the
+cluster-version-operator using static manifests that include the
+resource requests. Those operators then deploy other controllers or
+workloads dynamically. We need to account for both sets of components.
+
+### Deployments managed by the cluster-version-operator
+
+To start, we will [reduce the default
+requests](https://bugzilla.redhat.com/show_bug.cgi?id=1920159) for
+cluster operators deployed with static manifests. By re-tuning those
+settings by hand, we may be able to avoid implementing a more complex
+system to change them dynamically within the cluster.
+
+### Deployments managed by cluster operators
+
+Tuning for other workloads should be managed via an API, so that
+operators can adjust their settings correctly. We do not want users to
+change the setting after building the cluster (at least for now), so
+the API can be a status field. As with other infrastructure-related
+settings, the field will be added to the
+`infrastructure.config.openshift.io` [informational
+resource](https://docs.openshift.com/container-platform/4.6/installing/install_config/customizations.html#informational-resources_customizations).
+
+The new `controlPlaneCPULimit` field will be a positive integer
+representing the whole number of [Kubernetes
+CPUs](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#meaning-of-cpu)
+into which the control plane must fit. Expressing the API this way
+makes it easy for the user to declare the requirements up front when
+installing the cluster, without having to understand the basis for the
+default settings and without having to calculate a scaling factor
+themselves.
+
+In order for operators to know how to use `controlPlaneCPULimit` to
+compute a scaling factor, they need to know the default number of CPUs
+assumed for a control plane that produced the default resource
+limit. Rather than encoding that information separately in every
+operator implementation, we should put it in one place so it can be
+managed consistently. Using a `const` defined in the library with the
+Infrastructure type makes it easy to upgrade operators if the value
+changes across releases. So, a new `DefaultControlPlaneCPUCount`
+constant will be added to the module with the Infrastructure type.
+
+If `controlPlaneCPULimit == 0`, operators should use their
+default configuration for their operands.
+
+If `0 < controlPlaneCPULimit < DefaultControlPlaneCPUCount`, operators
+should scale their operands down by around `(controlPlaneCPULimit /
+DefaultControlPlaneCPUCount)` or more, if appropriate. When scaling,
```

Comment:
One consideration is that there may not be a one size fits all scaling factor based on the use case. For Telco use cases, some operators are not required whereas they may be for other use cases. The console would be a good example. In the former case, we would want to be more aggressive with the scale down.

Reply:
Operators that can be disabled will have their own API to do that, and those changes would be covered by other enhancements. So for the purposes of this doc, assume we're talking only about how to scale the things we haven't disabled or re-configured in some application-specific way. I can update the introduction to clarify that.

The goal here was to describe a general API that would apply to everything running in the cluster. If we need finer control over scaling for some specific operators to handle special use cases, those would be application-specific APIs, like the ones to disable operators. I can update the non-goals section to talk about creating those in separate enhancements.

Reply:
I updated the goals and non-goals lists to clarify the scope based on this thread.